clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -0,0 +1,319 @@
+     */
+    private LinkedHashMap<TopicPartition, PartitionData> sessionParts = null;
+
+    public static class FetchRequestData {
+        /**
+         * The partitions to send in the fetch request.
+         */
+        private final Map<TopicPartition, PartitionData> toSend;
+
+        /**
+         * All of the partitions which exist in the fetch request session.
+         */
+        private final Map<TopicPartition, PartitionData> sessionParts;
+
+        /**
+         * The metadata to use in this fetch request.
+         */
+        private final FetchMetadata metadata;
+
+        FetchRequestData(Map<TopicPartition, PartitionData> toSend,
+                         Map<TopicPartition, PartitionData> sessionParts,
+                         FetchMetadata metadata) {
+            this.toSend = toSend;
+            this.sessionParts = sessionParts;
+            this.metadata = metadata;
+        }
+
+        /**
+         * Get the set of partitions to send in this fetch request.
+         */
+        public Map<TopicPartition, PartitionData> toSend() {
+            return toSend;
+        }
+
+        /**
+         * Get the full set of partitions involved in this fetch request.
+         */
+        public Map<TopicPartition, PartitionData> sessionParts() {
+            return sessionParts;
+        }
+
+        public FetchMetadata metadata() {
+            return metadata;
+        }
+    }
+
+    public class Builder {
+        private LinkedHashMap<TopicPartition, PartitionData> next = new LinkedHashMap<>();
+
+        /**
+         * Mark that we want to remove a partition from the incremental fetch session.
+         */
+        public void remove(TopicPartition topicPartition) {
+            next.remove(topicPartition);
+        }
+
+        /**
+         * Mark the data which we want from this partition on the upcoming fetch.
+         */
+        public void add(TopicPartition topicPartition, PartitionData data) {
+            next.put(topicPartition, data);
+        }
+
+        public FetchRequestData build() {
+            String fullRequestReason = null;
+            if (sessionParts == null) {
+                fullRequestReason = "Initializing new session";
+            } else if (prevMetadata.type() != INCREMENTAL) {
+                fullRequestReason = "Previous response was not incremental";
+            }
+            if (fullRequestReason != null) {
+                FetchMetadata metadata = new FetchMetadata(FULL,
+                    prevMetadata.sessionId(), prevMetadata.sessionEpoch());
+                log.trace("{} for node {}. Built fetch {} with {} partition(s).",
+                    fullRequestReason, node, metadata, next.size());
+                sessionParts = next;
+                next = null;
+                Map<TopicPartition, PartitionData> toSend =
+                    Collections.unmodifiableMap(new LinkedHashMap<>(sessionParts));
```

Comment:
The ordering should be maintained, since `FetchSessionHandler#Builder#next` is a `LinkedHashMap`.  I guess there should be a comment about this in the code, so that it's documented.